File: internal/layout/layout.go
package layout

// Kind is a CSS grid shape a Page's images get arranged into.
type Kind int

const (
	// Text is a page with no images — content just fills the screen.
	Text Kind = iota
	// Split is one image: two columns, text and image side by side. The
	// renderer decides which column is which using Page.ImagesFirst —
	// Classify only decides the shape, never the order.
	Split
	// Stack is two images stacked in one column, text filling the other.
	Stack
	// Bento3 is one large image plus two smaller ones, with text as a
	// card filling the remaining grid cell.
	Bento3
	// Bento4 is a 2x2 grid of images, with text as a header/caption
	// strip above it.
	Bento4
)

// MaxPerPage is the most images Classify ever designs a shape for. A page
// with more images than this is not this package's problem to solve —
// deciding to split it into continuation pages happens in the renderer,
// which is the only package that depends on both markdown (to see the
// Page) and layout (to know this limit). Keeping the split out of here
// keeps Classify a pure function of a count, and keeps the Parser package
// from ever needing to import layout at all.
const MaxPerPage = 4

// Classify maps an image count to a grid shape. Counts above MaxPerPage
// clamp to Bento4 — the renderer is expected to never actually call
// Classify with more than MaxPerPage in the first place, once it's
// chunking overflow pages itself.
func Classify(n int) Kind {
	switch {
	case n <= 0:
		return Text
	case n == 1:
		return Split
	case n == 2:
		return Stack
	case n == 3:
		return Bento3
	default:
		return Bento4
	}
}

// String names a Kind for use as a CSS class or template output, e.g.
// class="layout-{{.String}}".
func (k Kind) String() string {
	switch k {
	case Text:
		return "text"
	case Split:
		return "split"
	case Stack:
		return "stack"
	case Bento3:
		return "bento-3"
	case Bento4:
		return "bento-4"
	default:
		return "text"
	}
}

File: extension/src/deck/layout.ts
// Port of internal/layout/layout.go. Same shapes, same names, same clamping
// behaviour — layouts.css keys off these strings, and that stylesheet is
// shared verbatim with the CLI, so the names are not free to drift.

export type LayoutKind = "text" | "split" | "stack" | "bento-3" | "bento-4";

/**
 * The most images a single screen ever gets a shape for. A page with more than
 * this is not this module's problem: splitting it into continuation screens
 * happens in render.ts, the only place that knows about both a Page and this
 * limit. Keeping the split out of here keeps classify a pure function of a
 * count.
 */
export const MAX_PER_PAGE = 4;

/**
 * Maps an image count to a grid shape:
 *   0 — text only, content fills the screen
 *   1 — split: two columns, text and image side by side
 *   2 — stack: two images in one column, text filling the other
 *   3 — bento-3: one large image plus two small, text as a card
 *   4+ — bento-4: a 2x2 image grid with text as a header strip
 *
 * Counts above MAX_PER_PAGE clamp to bento-4; the renderer is expected to
 * chunk overflow into continuation screens before it ever gets here.
 *
 * classify only decides the shape, never the order — Page.imagesFirst decides
 * which column is which.
 */
export function classify(n: number): LayoutKind {
  if (n <= 0) {
    return "text";
  }
  if (n === 1) {
    return "split";
  }
  if (n === 2) {
    return "stack";
  }
  if (n === 3) {
    return "bento-3";
  }
  return "bento-4";
}
